ard_controller/ard_controller.ino

```diff
@@ -15,45 +25,274 @@ struct JointAngleStruct {
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_ELBOW;
+        }
+        break;
+      case READ_R_ELBOW:
+        temporary_packet_data.right_elbow += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_CHECKSUM;
+        }
+        break;
+      case READ_CHECKSUM:
+        temporary_packet_data.checksum = received_data;
+
+        // calculated checksum must match received checksum; otherwise there's an error
+        if (temporary_packet_data.checksum == 0xFF - (calculated_checksum % 256)) {
+          // good packet
+          calculated_checksum = 0x00;
+          preamble_counter = PREAMBLE_LENGTH;
+
+          temporary_packet_data.checksum_error = false;
+          temporary_packet_data.packet_available = true;
+
+          validated_packet_data.left_hip = temporary_packet_data.left_hip;
+          validated_packet_data.left_knee = temporary_packet_data.left_knee;
+          validated_packet_data.right_hip = temporary_packet_data.right_hip;
+          validated_packet_data.right_knee = temporary_packet_data.right_knee;
+          validated_packet_data.left_shoulder = temporary_packet_data.left_shoulder;
+          validated_packet_data.left_elbow = temporary_packet_data.left_elbow;
+          validated_packet_data.right_shoulder = temporary_packet_data.right_shoulder;
+          validated_packet_data.right_elbow = temporary_packet_data.right_elbow;
+          validated_packet_data.checksum = temporary_packet_data.checksum;
+          validated_packet_data.checksum_error = temporary_packet_data.checksum_error;
+          validated_packet_data.packet_available = temporary_packet_data.packet_available;
+          validated_packet_data.data_request = temporary_packet_data.data_request;
+
+          sr_state = INIT;
+        }
+        else {
+          temporary_packet_data.checksum_error = true;
+          temporary_packet_data.packet_available = false;
+          sr_state = INIT;
+        }
+        break;
+      default:
+        sr_state = INIT;
+        break;
+      }
   }
 }
 
+void sensor_data_response() {
+  int raw_sum = 0;
+  int checksum;
+  long angle, goal;
+  int b1, b2;
+
+  // send preamble
+  for (int i = 0; i < PREAMBLE_LENGTH; i++) {
+    Serial1.write((byte)255);
+  }
+
+  // send imu and joint angle data
+  for (int i = 0; i < NUM_JOINTS + NUM_IMU_AXES; i++) {
+    angle = random(360);
+    raw_sum += angle;
+    b1 = angle / 256 > 0 ? 255 : angle;
+    b2 = angle / 256 > 0 ? angle % 255 : 0;
+
+    Serial1.write((byte)b1);
+    Serial1.write((byte)b2);
+  }
+
+  // send at_goal flag
+  goal = random(2);
+  raw_sum += goal;
+  Serial1.write((byte)goal);
+
+  // send checksum
+  checksum = 255 - raw_sum % 256;
```

Comment:
idk how useful this is to implement on the arduino side, but would it be possible to have something like a `MessageBuilder` class? Figure that every single message b/w the arduino & raspi will need to have a preamble as well as a checksum. If we had something like a `QueueMessage()` method or something like that, then `raw_sum` could automatically be tracked.

Then, having a `send()` method is trivial, especially cause the preamble and the checksum will be automatically computed.

Reply:
We could definitely do that, but I would think of that path as optional for the moment. The way I see it, the service response will be the only time the Arduino ever sends data through the serial port (unless we decide to implement a heartbeat for any reason). Since that's the reality, the only reason for creating an external class would be for organization. What would probably be a better idea to do (later on not right now) is to have a class dedicated to serial comms in general that would handle both incoming and outgoing data similar to the serial interface node on the ROS side.

Reply:
yea also realized that this the arduino and not the ros code--does arduino even have classes lol? yeah, I'd say we can move the message builder to a python package down the line.